### PR TITLE
Extract validation logic for Route into RouteValidator

### PR DESCRIPTION
**Pull request by nathancoleman**
### Changes proposed in this PR:
Analogous to #323.

This is a portion of the refactor changes from #163. The end result of this PR is that the logic for validating `xRoutes` is abstracted out into the `RouteValidator`. The route then proxies through to `validator.Validate(...)` rather than containing all of the validation logic itself.

### How I've tested this PR:
🤖 tests pass

### Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > Run `make changelog-entry` for guidance in authoring a changelog entry, and
  > commit the resulting file, which should have a name matching your PR number.
  > Entries should use imperative present tense (e.g. Add support for...)

### CODE REVIEW

File: internal/k8s/reconciler/validator/route.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+// all kubernetes routes implement the following two interfaces
+type Route interface {
+	client.Object
+	schema.ObjectKind
+}
+
+// RouteValidator is responsible for taking a provided v1alpha2.HTTPRoute or
+// v1alpha2.TCPRoute and deriving a state.RouteState from it. Ultimately, this
+// RouteState is what makes up the Status on the Kubernetes Route resource.
```
Comment (andrewstucki) [inline internal/k8s/reconciler/validator/route.go:26]:
```suggestion
// RouteState is what makes up the Status on the Kubernetes Route resource
// and holds any intermediate state resolved from remote object references
// (i.e. resolved Consul services from `backendRefs` values).
```

**Review by andrewstucki — APPROVED** [review]
LGTM once tests run/pass.

File: internal/k8s/reconciler/validator/route.go (hunk 1 above)
Comment (andrewstucki) [reply internal/k8s/reconciler/validator/route.go:26]: Also, looks like we have a mismatch for `v1beta1.Gateway` in the above docstring v. `v1alpha2.HTTP/TCPRoute` here.

File: internal/k8s/reconciler/validator/utils.go [hunk 2]
```diff
@@ -11,35 +11,6 @@ import (
 	"github.com/hashicorp/consul-api-gateway/internal/k8s/gatewayclient"
 )
 
-// gatewayAllowedForSecretRef determines whether the gateway is allowed
-// for the secret either by being in the same namespace or by having
-// an applicable ReferenceGrant in the same namespace as the secret.
-func gatewayAllowedForSecretRef(ctx context.Context, gateway *gwv1beta1.Gateway, secretRef gwv1beta1.SecretObjectReference, c gatewayclient.Client) (bool, error) {
```
Comment (nathancoleman) [inline internal/k8s/reconciler/validator/utils.go:17]: Moved into `validator/gateway.go` to maintain symmetrical package structure

**Pull request merged**